File: src/br/ufrj/ppgi/parser/ElementoXML.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package br.ufrj.ppgi.parser;

import java.lang.reflect.Array;
import java.util.ArrayList;

/**
 *
 * @author Rafael Pinheiro
 */
public class ElementoXML {
    
	public enum TipoElemento{
    	FILHO,
    	ATRIBUTO,
    	TEXTO    
    }
	
	private Integer id;
    private String nome;
    private ArrayList<ElementoXML> filhos = new ArrayList<ElementoXML>();
    private ArrayList<String> conteudoTexto = new ArrayList<String>();
    private ElementoXML pai;
    private TipoElemento tipo;
    private boolean bElementoImpresso = false;
    
    
    public void setElementoImpresso(boolean bImpresso){
    	bElementoImpresso = bImpresso;
    }
    
    public boolean getElementoImpresso(){
    	return bElementoImpresso;
    }
    
    
    public void setTipo(TipoElemento tipo){
    	this.tipo = tipo;
    }
    
    public TipoElemento getTipo(){
    	return this.tipo;
    }
    
    public void setId(Integer id){
    	this.id = id;
    }
    
    public Integer getId(){
    	return id;
    }
    
    public void setNome(String nome){
    	this.nome = nome;
    }
    
    public String getNome(){
    	return nome;
    }
    
    public void adicionarFilho(ElementoXML novoFilho){
    	this.filhos.add(novoFilho);
    }
    
    public ElementoXML getPai(){
    	return pai;
    }
    
    public void setPai(ElementoXML pai){
    	this.pai = pai;
    }
    
    public void setConteudo(String conteudo){
    	this.conteudoTexto.add(conteudo);
    }
    
    public ArrayList<String> getConteudoTexto(){
    	return this.conteudoTexto;
    }
    
    public String getConteudo(){
    	StringBuilder conteudo = new StringBuilder();
    	if (conteudoTexto.size() == 0){
    		int a = 0;conteudo.append("");
    	}
    	else
    	{
	    	for( int i = 0; i < conteudoTexto.size(); i++ ){
	    		conteudo.append(conteudoTexto.get(i));
	    	}
    	}
    	return conteudo.toString();
    }
    
    public Boolean temSomenteQuebraDeLinha(){
    	if ( (getConteudo().indexOf("\n") > -1 && getConteudo().indexOf("\t") > -1) ||
    		(getConteudo().indexOf("\n") > -1 && conteudoTexto.size() == 1))
    		return true;
    	
    	return false;
    }
    
    public ArrayList<ElementoXML> getFilhos(){
    	return this.filhos;
    
    }
    
    public Boolean ehElementoMisto(){
    	Boolean bEncontrouTexto = false;
    	Boolean bEncontrouElemento = false;
		for( int i = 0; i < filhos.size(); i++ ){
			if ( filhos.get(i).getTipo() == TipoElemento.TEXTO )
				bEncontrouTexto = true;
			
			if ( filhos.get(i).getTipo() == TipoElemento.FILHO )
				bEncontrouElemento = true;
				
		}
		
		if ( bEncontrouElemento && bEncontrouTexto )
			return true;
    	
    	return false;
    }
    
    public boolean naoPossuiElementoTexto(){
    	boolean bPossuiElementoTexto = false;
    	for( int i = 0; i < filhos.size(); i++ ){
			if ( filhos.get(i).getTipo() == TipoElemento.TEXTO ){
				bPossuiElementoTexto = true;
				break;
			}			
		}
    	return bPossuiElementoTexto;
    }
}

File: C++ Projects/SaxParser/ElementoXML.cpp
#include "stdafx.h"
#include "ElementoXML.h"


ElementoXML::ElementoXML(void)
{
	pPai = 0;
	id = -1;
	idPai = -1;
	idAvo = -1;
	bElementoImpresso = false;
	listaFilhos.clear();
	listaConteudo.clear();
}


ElementoXML::~ElementoXML(void)
{
	listaConteudo.clear();
	list<ElementoXML*>::iterator itLista;
	for(itLista = listaFilhos.begin(); itLista != listaFilhos.end();)
	{
		delete (*itLista);
		(*itLista) = 0;
		itLista = listaFilhos.erase(itLista);
	}
	//listaFilhos.clear();

	pPai = 0;
	/*if ( pPai != NULL )
		delete pPai;*/
}

void ElementoXML::adicionarFilho(ElementoXML* filho)
{
	listaFilhos.push_back(filho);
}
void ElementoXML::setId(int novoId)
{
	id = novoId;
}

void ElementoXML::setNome(string& novoNome)
{
	nome = novoNome;
}

void ElementoXML::setTipo(ElementoXML::Tipo novoTipo)
{
	tipo = novoTipo;
}

void ElementoXML::setConteudo(string& novoConteudo)
{
	listaConteudo.push_back(novoConteudo);
}

void ElementoXML::setPai(ElementoXML* novoPai)
{
	pPai = novoPai;
}

int ElementoXML::getId()
{
	return id;
}

string ElementoXML::getNome()
{
	return nome;
}

ElementoXML::Tipo ElementoXML::getTipo()
{
	return tipo;
}

string ElementoXML::getConteudo()
{
	string conteudo;
	list<string>::iterator itLista;
	list<string>::iterator itTempLista = listaConteudo.end();
	itTempLista--;
	for(itLista = listaConteudo.begin() ; itLista != listaConteudo.end(); itLista++)
	{
		conteudo += (*itLista);
		if ( itLista != itTempLista)
			conteudo += " ";

	}

	return conteudo;
}

ElementoXML* ElementoXML::getPai()
{
	return pPai;
}

const list<ElementoXML*> ElementoXML::getFilhos()
{
	return listaFilhos;
}

bool ElementoXML::ehElementoMisto()
{
	if ( getFilhos().size() == 0 )
		return false;

	bool bTemElementoTexto = false;
	bool bTemElementoFilho = false;
	list<ElementoXML*>::iterator itListaFilhos;
	for( itListaFilhos = listaFilhos.begin(); itListaFilhos != listaFilhos.end(); itListaFilhos++ )
	{
		if( (*itListaFilhos)->getTipo() == ElementoXML::Tipo::TEXTO )
			bTemElementoTexto = true;

		if( (*itListaFilhos)->getTipo() ==  ElementoXML::FILHO )
			bTemElementoFilho = true;
	}

	return bTemElementoTexto && bTemElementoFilho;
}

void ElementoXML::setElementoImpresso(bool bImpresso)
{
	bElementoImpresso = bImpresso;
}

bool ElementoXML::ehElementoImpresso()
{
	return bElementoImpresso;
}

bool ElementoXML::naoPossuiElementoTexto()
{
    bool bPossuiElementoTexto = false;
	list<ElementoXML*>::iterator itListaFilhos;
    for( itListaFilhos = listaFilhos.begin(); itListaFilhos != listaFilhos.end(); itListaFilhos++ )
	{
		if ( (*itListaFilhos)->getTipo() == ElementoXML::Tipo::TEXTO )
		{
			bPossuiElementoTexto = true;
			break;
		}			
	}
    return bPossuiElementoTexto;
}


bool ElementoXML::possuiElementoFilho()
{
    bool bPossuiElementoTexto = false;
	list<ElementoXML*>::iterator itListaFilhos;
    for( itListaFilhos = listaFilhos.begin(); itListaFilhos != listaFilhos.end(); itListaFilhos++ )
	{
		if ( (*itListaFilhos)->getTipo() == ElementoXML::Tipo::FILHO ) 
		{
			bPossuiElementoTexto = true;
			break;
		}			
	}
    return bPossuiElementoTexto;
}
